src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScannerTest.java

```diff
@@ -1020,11 +1022,73 @@ public void inProgressParallelInParallel() throws Exception {
         SemaphoreStep.waitForStart("innerA/1", b);
-        ForkScanner scanner = new ForkScanner();
         sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
+        assertBranchOrder(b.getExecution(), "innerB", "innerA", "outerC", "outerB", "outerA");
         SemaphoreStep.success("outerA/1", null);
         SemaphoreStep.success("innerA/1", null);
         r.assertBuildStatusSuccess(r.waitForCompletion(b));
         sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
+        assertBranchOrder(b.getExecution(), "innerB", "innerA", "outerC", "outerB", "outerA");
+    }
+
+    @Test
+    public void inProgressParallelInParallelOneNestedBranch() throws Exception {
+        WorkflowJob p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                "parallel(\n" +                   // 3
+                "  'outerA': {\n" +               // 5
+                "    parallel(\n" +               // 7
+                "      'innerA': {\n" +           // 8
+                "        semaphore('innerA')\n" + // 10
+                "      }\n" +                     // 11
+                "    )\n" +                       // 12
+                "  },\n" +                        // 13
+                "  'outerB': {\n" +               // 6
+                "  }\n" +                         // 9
+                ")", true));                      // 14
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("innerA/1", b);
+        sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
+        // TODO: We would like for this order to match the order of the completed build, but that requires reverting
+        // https://github.com/jenkinsci/workflow-api-plugin/pull/287, which would reintroduce a much more severe problem.
+        assertBranchOrder(b.getExecution(), "innerA", "outerA", "outerB");
+        SemaphoreStep.success("innerA/1", null);
+        r.assertBuildStatusSuccess(r.waitForCompletion(b));
+        sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
+        assertBranchOrder(b.getExecution(), "outerB", "innerA", "outerA");
+    }
+
+    @Test
+    public void parallelBranchOrdering() throws Exception {
+        WorkflowJob p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                "parallel(\n" +
+                "  '1': { echo '1' },\n" +
+                "  '2': { semaphore('2') },\n" +
+                "  '3': { echo '3' },\n" +
+                "  '4': { semaphore('4') },\n" +
+                "  '5': { echo '5' },\n" +
+                ")", true));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("2/1", b);
+        SemaphoreStep.waitForStart("4/1", b);
+        sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
+        assertBranchOrder(b.getExecution(), "5", "4", "3", "2", "1");
```

Comment:
Without this PR, this order is pretty unusual, I think it's 4 2 1 3 5, or maybe it was 1 3 5 4 2, either way it was not a trivial case of the list being backwards or something like that.